infra/src/main/kotlin/fi/fta/geoviite/infra/linking/SwitchLinkingService.kt

```diff
@@ -1072,16 +1074,124 @@ class SwitchLinkingService @Autowired constructor(
 
         switchJointsByLocationTrack.forEach { (locationTrackId, switchJoints) ->
             val (locationTrack, alignment) = locationTrackService.getWithAlignmentOrThrow(DRAFT, locationTrackId)
+
+            val switchJointsWithSlightlyOverlappingSegmentsSnapped = switchJoints.map { switchLinkingJoint ->
+                switchLinkingJoint.copy(
+                    segments = switchLinkingJoint.segments.map { switchLinkingSegment ->
+                        val referencedLayoutSegment = alignment.segments[switchLinkingSegment.segmentIndex]
+
+                        val layoutSegmentHasExistingSwitch = referencedLayoutSegment.switchId != null
+                        val layoutSegmentSwitchReferencesDifferentSwitch =
+                            layoutSegmentHasExistingSwitch && referencedLayoutSegment.switchId != linkingParameters.layoutSwitchId
+
+                        if (layoutSegmentSwitchReferencesDifferentSwitch) {
+                            tryToSnapOverlappingSwitchSegmentToNearbySegment(alignment, switchLinkingSegment)
+                        } else {
+                            switchLinkingSegment
+                        }
+                    }
+                )
+            }
+
             val updatedAlignment = updateAlignmentSegmentsWithSwitchLinking(
                 alignment = alignment,
                 layoutSwitchId = linkingParameters.layoutSwitchId,
-                matchingJoints = switchJoints,
+                matchingJoints = switchJointsWithSlightlyOverlappingSegmentsSnapped,
             )
             val updatedLocationTrack = locationTrackService.updateTopology(locationTrack, updatedAlignment)
             locationTrackService.saveDraft(updatedLocationTrack, updatedAlignment)
         }
     }
 
+    private fun findExistingSwitchEdgeSegmentWithSwitchFreeAdjacentSegment(
+        existingSwitchId: IntId<TrackLayoutSwitch>,
+        layoutSegments: List<LayoutSegment>,
+        searchIndexRange: IntProgression,
+    ): IndexedValue<LayoutSegment>? {
+
+        if (searchIndexRange.any { index -> index !in layoutSegments.indices }) {
+            logger.info("Invalid searchIndexRange: $searchIndexRange contains indices outside of layoutSegments (${layoutSegments.indices})")
```

Comment:
Tälle ei liene mitään järkisyytä miksi tätä pitäisi jostain kutsua ohi-indekseillä? Jos näin, niin kannattaa vain heittää poikkeus - saadaan bugi kiinni nopeammin ja oikeassa paikassa. Vastaavia on muualla tehty `require()` ehtoina.

Reply:
Jos kaikki nuo null-tilanteet on oikeasti koodivirheitä niin paluuarvokin voisi olla not-null ja lopussa heittää jos haettua ei löytynyt. Vai onko tilanteita joissa on oikeasti eduksi kysellä sellaisen vaihteen perään jota raiteella ei ole?

Reply:
Näihin haaroihin pystyi kyllä päätymään suht normaaleissa tilanteissa. Esimerkiksi kutsujahan ei varmistanut aloitusindeksin ja lopetusindeksin olevan pienempää/suurempaa riippuen suunnasta, mutta täältä palautui vain nulli -- joka toimi.

Muutin näihin indeksitarkistuksiin requiret tähän funktioon ja kutsujaa varmistamaan, ettei virheellisellä searchIndexRangella pitäisi tätä funktiota enää kutsua. Kysymyshän on vähän siitä, että mikä funktio on vastuussa asioiden varmistamisesta/päättää mitä funktiota kutsutaan.

Nuo indeksitarkistukset päätyivät myös tuosta jokaisen elementin tarkistuksesta yksinkertaisempaan muotoon ( O(n) -> O(1) ). 

Sen sijaan tuo luupin sisäinen tarkistus ja lopun nullin palautus voivat kyllä sattua ainakin linkitettäessä uutta vaihdetta kahden aiemmin linkitetyn vaihteen "päälle" (ks. vastaus Juusen kysymykseen).

Tällä funktiollahan on ainakin kaksi roolia (etsii vaihteen alku/loppusegmentin JA tarkistaa onko sen viereinen segmentti vaihteesta vapaa vai ei). Voisipa siis olla vielä järkevää jakaa erillisiksi.